master/internal/rm/multirm/multirm.go

```diff
@@ -0,0 +1,435 @@
+)
+
+// ErrRPNotDefined returns a detailed error if a resource manager/pool combination isn't found in the MultiRMRouter map.
+func ErrRPNotDefined(rm, rp string) error {
+	return fmt.Errorf("resource pool %s not found in resource manager %s", rp, rm)
+}
+
+// MultiRMRouter tracks all resource managers in the system.
+type MultiRMRouter struct {
+	mu     sync.Mutex
+	rms    map[string]rm.ResourceManager
+	syslog *logrus.Entry
+}
+
+// New creates a new Multi-RM router with the master set-up.
+func New(
+	db *db.PgDB,
+	echo *echo.Echo,
+	rmConfigs []*config.ResourceManagerWithPoolsConfig,
+	tcd *model.TaskContainerDefaultsConfig,
+	opts *aproto.MasterSetAgentOptions,
+	cert *tls.Certificate,
+) rm.ResourceManager {
+	if len(rmConfigs) <= 1 {
+		config := rmConfigs[0]
+		switch {
+		case config.ResourceManager.AgentRM != nil:
+			return agentrm.New(db, echo, config, opts, cert)
+		case config.ResourceManager.KubernetesRM != nil:
+			return kubernetesrm.New(db, config, tcd, opts, cert)
+		default:
+			panic("no expected resource manager config is defined")
+		}
+	}
+
+	rms := map[string]rm.ResourceManager{}
+	for _, cfg := range rmConfigs {
+		c := cfg
+		switch {
+		case c.ResourceManager.AgentRM != nil:
+			// TODO (multirm): what to do if Name is ""?
+			rms[c.ResourceManager.AgentRM.Name] = agentrm.New(db, echo, c, opts, cert)
+		case c.ResourceManager.KubernetesRM != nil:
+			// TODO (multirm): ditto.
+			rms[c.ResourceManager.KubernetesRM.Name] = kubernetesrm.New(db, c, tcd, opts, cert)
+		default:
+			panic("no expected resource manager config is defined")
+		}
+	}
+	return &MultiRMRouter{
+		rms:    rms,
+		syslog: logrus.WithField("component", "resource-router"),
+	}
+}
+
+// GetAllocationSummaries returns the allocation summaries for all resource pools across all resource managers.
+func (m *MultiRMRouter) GetAllocationSummaries() (
+	map[model.AllocationID]sproto.AllocationSummary,
+	error,
+) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	var eg errgroup.Group
+	var mu sync.Mutex
+	res := map[model.AllocationID]sproto.AllocationSummary{}
+
+	for _, rm := range m.rms {
+		r := rm
+		eg.Go(func() error {
+			res1, err := r.GetAllocationSummaries()
+			if err != nil {
+				return err
+			}
+
+			// TODO (multirm): I know this looks ugly, but I couldn't
+			// figure out how to *not* panic if I send this to a channel.
+			mu.Lock()
+			// Note: this is a built-in function that will rewrite on duplicates.
+			maps.Copy(res, res1)
+			mu.Unlock()
+			return nil
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	return res, nil
+}
+
+// Allocate routes an AllocateRequest to the specified RM.
+func (m *MultiRMRouter) Allocate(rmName string, req sproto.AllocateRequest) (*sproto.ResourcesSubscription, error) {
+	rmForRp, ok := m.rms[rmName]
+	if !ok {
+		return nil, ErrRPNotDefined(rmName, req.ResourcePool)
+	}
+	return rmForRp.Allocate(rmName, req)
+}
+
+// Release routes an allocation release request.
```

Comment:
Does this function not need to grab the mutex? (I'm a big fan of not over-locking, I guess I'm wondering if there's themes around when we usually use a mutex?)

Reply:
Ah we don't -- and it doesn't make sense in the flow for the resource manager interface anyway. at the moment, I don't think we expect to add Resource Managers past the initial config/master setup